pkg/controller/elasticsearch/driver/upgrade.go

```diff
@@ -236,8 +246,27 @@ func doFlush(ctx context.Context, es esv1.Elasticsearch, esClient esclient.Clien
 		return err
 	}
 }
+func (d *defaultDriver) maybeCompleteNodeUpgrades(
+	ctx context.Context,
+	esClient esclient.Client,
+	esState ESState,
+	nodeShutdown *shutdown.NodeShutdown,
+) *reconciler.Results {
+	// we still have to enable shard allocation in cases where we just upgraded from
+	// a version that did not support node shutdown to a supported version.
```

Comment:
I was wondering if we can be smarter here, and only enable shard allocation if necessary. But I'm not sure there is an easy way to know that we have just restarted the last node that did not support node shutdown.
It would require to store the previous version (as in `ElasticsearchStatus.Version`) or persist somewhere the fact that shards allocation has been disabled on purpose by the operator.

Reply:
We could track state in an annotation on a best-effort basis. Once we migrated a cluster fully to node shutdown we would set the annotation and when it is present we would not try to enable shard allocations anymore.

Reply:
Let's keep it simple for now, I'm not convinced that it really deserves the complexity of handling an additional annotation.